src/linear_algebra/matrix/circulant.lean

```diff
@@ -0,0 +1,246 @@
+/-
+Copyright (c) 2021 Lu-Ming Zhang. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Lu-Ming Zhang
+-/
+import linear_algebra.matrix.symmetric
+import data.polynomial.monomial
+import data.matrix.pequiv
+import data.equiv.fin
+
+/-!
+# Circulant matrices
+
+This file contains the definition and basic results about circulant matrices.
+
+## Main results
+
+- `matrix.cir`: introduce the definition of a circulant matrix
+                generated by a given vector `v : I → α`.
+
+## Implementation notes
+
+`fin.foo` is the `fin n` version of `foo`.
+Namely, the index type of the circulant matrices in discussion is `fin n`.
+
+## Tags
+
+cir, matrix
+-/
+
+variables {α β I J R : Type*} {n : ℕ}
+
+namespace matrix
+
+open function
+open_locale matrix big_operators
+
+/-- Given the condition `[has_sub I]` and a vector `v : I → α`,
+    we define `cir v` to be the circulant matrix generated by `v` of type `matrix I I α`. -/
+def cir [has_sub I] (v : I → α) : matrix I I α
+| i j := v (i - j)
+
+lemma cir_col_zero_eq [add_group I] (v : I → α) :
+  (λ i, (cir v) i 0) = v :=
+by ext; simp [cir]
+
+lemma cir_ext_iff [add_group I] {v w : I → α} :
+  cir v = cir w ↔ v = w :=
+begin
+  split,
+  { intro h, rw [← cir_col_zero_eq v, ← cir_col_zero_eq w, h] },
```

Comment:
If you use this part of the proof to prove `cir_injective` (that is, swap the order you prove them in), then you can prove `cir_inj` as `cir_injective.eq_iff`

Reply:
I'll take a look.